service/history/archival/service.go

```diff
@@ -0,0 +1,265 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination service_mock.go
+
+package archival
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.uber.org/multierr"
+
+	archiverspb "go.temporal.io/server/api/archiver/v1"
+	carchiver "go.temporal.io/server/common/archiver"
+	"go.temporal.io/server/common/archiver/provider"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+type (
+	Request struct {
+		ShardID     int32
+		NamespaceID string
+		Namespace   string
+		WorkflowID  string
+		RunID       string
+
+		// history archival
+		BranchToken          []byte
+		NextEventID          int64
+		CloseFailoverVersion int64
+		HistoryURI           string
+
+		// visibility archival
+		WorkflowTypeName string
+		StartTime        time.Time
+		ExecutionTime    time.Time
+		CloseTime        time.Time
+		Status           enumspb.WorkflowExecutionStatus
+		HistoryLength    int64
+		Memo             *commonpb.Memo
+		SearchAttributes *commonpb.SearchAttributes
+		VisibilityURI    string
+
+		// archival targets: history and/or visibility
+		Targets       []Target
+		CallerService string
+	}
+
+	Target string
+
+	Response struct {
+	}
+
+	// Service is used to archive workflow history and visibility data. If a target fails, it returns an error, unlike
+	// Client, which will try to signal an archival workflow whenever an error occurrs.
+	Service interface {
```

Comment:
nit: we have four services(frontend, history matching, worker). What about just keep it as archiver?